src/cargo/util/queue.rs

```diff
@@ -0,0 +1,86 @@
+use std::collections::VecDeque;
+use std::sync::{Condvar, Mutex};
+use std::time::{Duration, Instant};
+
+/// A simple, threadsafe, queue of items of type `T`
+///
+/// This is a sort of channel where any thread can push to a queue and any
+/// thread can pop from a queue.
+///
+/// This supports both bounded and unbounded operations. `push` will never block,
+/// and allows the queue to grow without bounds. `push_bounded` will block if the
+/// queue is over capacity, and will resume once there is enough capacity.
+pub struct Queue<T> {
+    state: Mutex<State<T>>,
+    popper_cv: Condvar,
+    bounded_cv: Condvar,
+    bound: usize,
+}
+
+struct State<T> {
+    items: VecDeque<T>,
+}
+
+impl<T> Queue<T> {
+    pub fn new(bound: usize) -> Queue<T> {
+        Queue {
+            state: Mutex::new(State {
+                items: VecDeque::new(),
+            }),
+            popper_cv: Condvar::new(),
+            bounded_cv: Condvar::new(),
+            bound,
+        }
+    }
+
+    pub fn push(&self, item: T) {
+        self.state.lock().unwrap().items.push_back(item);
+        self.popper_cv.notify_one();
+    }
+
+    /// Pushes an item onto the queue, blocking if the queue is full.
+    pub fn push_bounded(&self, item: T) {
+        let mut state = self.state.lock().unwrap();
+        loop {
```

Comment:
This might be able to make use of the nifty [`wait_until` method](https://doc.rust-lang.org/std/sync/struct.Condvar.html#method.wait_until):

```rust
let state = self.bounded_cv.wait_until(state, |s| s.items.len() < self.bound).unwrap();
```

Reply:
Didn't know that existed!